custom_components/waste_collection_schedule/waste_collection_schedule/source/thanet_gov_uk.py

```
@@ -0,0 +1,102 @@
+    "en": {
+        "uprn": "Unique Property Reference Number (UPRN)",
+        "street_address": "House name OR Number and Street name",
+        "postcode": "Postcode",
+    }
+}
+
+
+class Source:
+    
+    header_text = { 'Accept-Language' :"en-GB,en;q=0.9,en-US;q=0.8",
+            'User-Agent': "Mozilla/5.0",
+            }
+    
+    
+    def __init__(self, uprn=None, postcode=None, street_address=None):
+        self._postcode = postcode
+        self._street_address = str(street_address).upper()
+        self._uprn = uprn
+
+    def fetch(self) -> List[Collection]:
+        if self._uprn is None:
+            self._uprn = self.get_uprn()
+
+        url = f"https://www.thanet.gov.uk/wp-content/mu-plugins/collection-day/incl/mu-collection-day-calls.php?pAddress={self._uprn}"
+        collections_json = requests.get(url, headers= self.header_text).json()
+        
+        entries = []
+        
+        for collection in collections_json:
+            for bin_type in TYPES:
+                if collection["type"] == bin_type:
+                    entries.append(
+                        Collection(
+                            date=datetime.strptime(collection['nextDate'][:10], "%d/%m/%Y").date(),
+                            t=TYPES[bin_type]['alias'],
+                            icon=TYPES[bin_type]['icon'],
+                        )
+                    )
+                    entries.append(
+                        Collection(
+                            date=datetime.strptime(collection['previousDate'][:10], "%d/%m/%Y").date(),
+                            t=TYPES[bin_type]['alias'],
+                            icon=TYPES[bin_type]['icon'],
+                        )
+                    )
+        return entries
+        
+    def get_uprn(self) -> str:
+        url = f"https://www.thanet.gov.uk/wp-content/mu-plugins/collection-day/incl/mu-collection-day-calls.php?searchAddress={requests.utils.quote(self._postcode)}"
```

Please use the `params` argument instead of using `requests.utils.quote`

I think it's a good idea to check if self._postcode and self._street_address is set. If no argument is set (which is possible with this __init__ arguments, the source throws `TypeError: quote_from_bytes() expected bytes` which is not really explanatory.

You could throw `SourceArgumentRequired` if `postcode` is missing and `SourceArgumentRequiredWithSuggestions` if `postcode` is set but not `street_address`

---

I think I've done this as described (a couple of days on from Hello World here so all of your points are really appreciated, I hope it's not too frustrating from your side. I've tried to learn by examples I have seen but these direct pointers are incredibly helpful).

---

It's not frustrating at all. I appreciate your work and am happy to help.